File: core/src/main/java/org/aresclient/ares/api/event/AresEventManager.java
package org.aresclient.ares.api.event;

import dev.tigr.simpleevents.EventManager;
import org.aresclient.ares.api.Ares;
import org.aresclient.ares.api.minecraft.util.Profiler;

public class AresEventManager extends EventManager {
    @Override
    public <T> T post(T event) {
        if(event instanceof AresEvent) {
            AresEvent aresEvent = (AresEvent) event;
            Profiler profiler = Ares.getMinecraft().getProfiler();
            profiler.push("ares_" + aresEvent.getName() + (aresEvent.getEra() != null ? "_" + aresEvent.getEra().name().toLowerCase() : ""));
            T out = super.post(event);
            profiler.pop();
            return out;
        } else return super.post(event);
    }
}

File: src/main/kotlin/org/aresclient/ares/api/events/AresEventManager.kt
package org.aresclient.ares.api.events

import dev.tigr.simpleevents.EventManager
import dev.tigr.simpleevents.event.Event
import org.aresclient.ares.api.Wrapper

enum class Era { BEFORE, AFTER }
abstract class AresEvent(val name: String, val era: Era? = null): Event(), Wrapper
class AresEventManager: EventManager(), Wrapper {

	override fun <T: Any?> post(event: T): T = when (event) {
		is AresEvent -> {
			profiler.push("ares_" + event.name.lowercase() + event.era?.let { "_" + it.name.lowercase() })
			val out = super.post(event)
			profiler.pop()
			out
		}
		else         -> {
			super.post(event)
		}
	}

}

